File: leetcode/daily_02.2022/7_389.cpp
#include <iostream>
#include <set>
#include <string>
#include <algorithm>

using namespace std;

// 389. Find the Difference

class Solution {
public:
    char findTheDifference(string s, string t) {
        multiset<char> tt(t.begin(), t.end());
        for(int i = 0; i < s.length(); i++) {
            tt.erase(tt.find(s[i]));
        }
        return *tt.upper_bound('A');
    }
    char findTheDifference2(string s, string t) {
        sort(s.begin(),s.end());
        sort(t.begin(),t.end());
        for(int i = 0; i < s.length(); i++)
        {
            if (s[i] != t[i]) return t[i];
        }
        return t[t.length()-1];
    }
};

int main()
{
    string s = "asbfkjabkalkanfklsafalnkfnddsaf";
    string t = "asbfkjabkalkafnfklsafalnkfnddsaf";
    Solution sol;
    cout << sol.findTheDifference(s, t) << "\n";
    cout << sol.findTheDifference2(s, t);
    return 0;
}

File: leetcode/daily_02.2022/7_389.py
class Solution:
    def findTheDifference(self, s: str, t: str) -> str:
        ans = 0
        for x in t: ans += ord(x)
        for x in s: ans -= ord(x)
        return chr(ans)
    
    def findTheDifference2(self, s: str, t: str) -> str:
        ans = 0
        # for x in s: ans ^= ord(x)
        # for x in t: ans ^= ord(x)
        for x in s+t: ans ^= ord(x)
        return chr(ans)
    
if __name__ == "__main__":
    s = Solution()
    st = "bakbsbremotebhrithnlebhtseltirsheritlerusob"
    t = "bakbsbremotebjhrithnlebhtseltirsheritlerusob"
    print(s.findTheDifference(st, t))
    print(s.findTheDifference2(st, t))